code/parachain/frame/cosmwasm/src/ibc.rs

```diff
@@ -15,32 +80,666 @@ impl<T: Config> Pallet<T> {
+		Ok(vm)
+	}
+
+	fn execute<I, M, V>(vm: &mut V, message: M) -> Result<I::Output, IbcError>
+	where
+		// unfortunately, here is reason
+		// 1. Rust fails to decided that VM is V (which it is)
+		// 2. also without unsafe rust cannot do struct which borrows mut 2 of its items (without
+		// callbacks) 3. so in order to reuse calls either need to build entrypoint.rs like wrappers for
+		// each call 4. but here just closed `cosmwasm_vm + pallet-cosmwasm` on functional level (helped
+		// rust type inference)
+		M: serde::Serialize,
+		I: Input + HasInfo,
+		I::Output: serde::de::DeserializeOwned + ReadLimit + DeserializeLimit,
+		V: cosmwasm_vm::vm::VM + ReadWriteMemory + Has<Env> + Has<MessageInfo> + VMBase,
+		<V as VMBase>::Error: std::fmt::Debug,
+		V::Pointer: for<'x> TryFrom<VmOutputOf<'x, V>, Error = VmErrorOf<V>>,
+		for<'x> Unit: TryFrom<VmOutputOf<'x, V>, Error = VmErrorOf<V>>,
+		for<'x> VmInputOf<'x, V>: TryFrom<AllocateInput<V::Pointer>, Error = VmErrorOf<V>>
+			+ TryFrom<DeallocateInput<V::Pointer>, Error = VmErrorOf<V>>
+			+ TryFrom<CosmwasmCallInput<'x, V::Pointer, I>, Error = VmErrorOf<V>>
+			+ TryFrom<CosmwasmCallWithoutInfoInput<'x, V::Pointer, I>, Error = VmErrorOf<V>>,
+		VmErrorOf<V>:
+			From<ReadableMemoryErrorOf<V>> + From<WritableMemoryErrorOf<V>> + From<ExecutorError>,
+	{
+		cosmwasm_call_serialize::<I, V, M>(vm, &message)
+			.map_err(|err| IbcError::implementation_specific(format!("{:?}", err)))
+	}
+
+	fn on_recv_packet_internal(
+		&self,
+		_output: &mut ModuleOutputBuilder,
+		packet: &ibc::core::ics04_channel::packet::Packet,
+		relayer: &pallet_ibc::Signer,
+	) -> Result<Vec<u8>, IbcError> {
+		let address = Self::port_to_address(&packet.destination_port)?;
+		let contract_info = Self::to_ibc_contract(&address)?;
+
+		let message = IbcPacketReceiveMsg {
+			packet: IbcPacket {
+				data: packet.data.clone().into(),
+				src: IbcEndpoint {
+					port_id: packet.source_port.to_string(),
+					channel_id: packet.source_channel.to_string(),
+				},
+				dest: IbcEndpoint {
+					port_id: packet.destination_port.to_string(),
+					channel_id: packet.destination_channel.to_string(),
+				},
+				sequence: packet.sequence.into(),
+				timeout: Err(IbcError::implementation_specific(
+					"https://app.clickup.com/t/39gjzw1".to_string(),
+				))?,
+			},
+			relayer: Addr::unchecked(relayer.to_string()),
+		};
+		let gas = Weight::MAX;
+		let mut vm = <Pallet<T>>::do_create_vm_shared(gas, InitialStorageMutability::ReadWrite);
+		let mut executor = Self::relayer_executor(&mut vm, address, contract_info)?;
+		let (data, _) = cosmwasm_system_entrypoint_serialize::<
+			IbcPacketReceive,
+			VM<T>,
+			IbcPacketReceiveMsg,
+		>(&mut executor, &message)
+		.map_err(|err| IbcError::implementation_specific(format!("{:?}", err)))?;
+		let _remaining = vm.gas.remaining();
+		Ok(data.expect("there is always data from contract; qed").0)
+	}
+}
+
+impl<T: Config + Send + Sync> IbcModule for Router<T> {
+	fn on_chan_open_init(
```

Comment:
This trait has been updated, the methods now provide access to the context to read channels and connections.
check the latest commit on the centauri repo.